File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AntiAFK.java
package me.zeroeightsix.kami.module.modules.misc;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.event.events.RenderEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.ColourUtils;
import me.zeroeightsix.kami.util.GeometryMasks;
import me.zeroeightsix.kami.util.KamiTessellator;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.network.play.client.CPacketAnimation;
import net.minecraft.network.play.server.SPacketChat;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import org.lwjgl.opengl.GL11;

import java.util.List;
import java.util.Objects;
import java.util.Random;

import static me.zeroeightsix.kami.util.ColourConverter.rgbToInt;
import static me.zeroeightsix.kami.util.MathsUtils.reverseNumber;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendChatMessage;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendServerMessage;
import static me.zeroeightsix.kami.util.VectorUtil.getBlockPositionsInArea;
import static me.zeroeightsix.kami.util.VectorUtil.getHighestTerrainPos;

/**
 * Created by 086 on 16/12/2017.
 * Updated by dominikaaaa on 21/04/20
 * TODO: Path finding to stay inside 1 chunk
 */
@Module.Info(
        name = "AntiAFK",
        category = Module.Category.MISC,
        description = "Prevents being kicked for AFK"
)
public class AntiAFK extends Module {

    private Setting<Integer> frequency = register(Settings.integerBuilder("Action Frequency").withMinimum(1).withMaximum(100).withValue(40).build());
    public Setting<Boolean> autoReply = register(Settings.b("AutoReply", true));
    private Setting<Mode> mode = register(Settings.enumBuilder(Mode.class).withName("Mode").withValue(Mode.TRADITIONAL).build());
    private Setting<Integer> size = register(Settings.integerBuilder("Chunk Size").withMinimum(1).withMaximum(10).withValue(1).build());
    private Setting<Boolean> swing = register(Settings.b("Swing", true));
    private Setting<Boolean> jump = register(Settings.b("Jump", true));
    private Setting<Boolean> turn = register(Settings.booleanBuilder("Turn").withValue(true).withVisibility(v -> mode.getValue().equals(Mode.TRADITIONAL)).build());

    private enum Mode { TRADITIONAL, CHUNK;}
    private Random random = new Random();

    private BlockPos pos1 = new BlockPos(0, 0, 0);
    private BlockPos pos2 = new BlockPos(0, 0, 0);
    private int[] cornerY = { 0, 0, 0, 0 }; // 22, 21, 12, 11
    private List<BlockPos> positions;

    public void onEnable() {
        if (mc.player == null) return;

        if (!mode.getValue().equals(Mode.CHUNK)) return;

        pos1.x = (int) mc.player.posX;
        pos1.y = (int) mc.player.posY;
        pos1.z = (int) mc.player.posZ;
        pos2.x = pos1.getX() + (16 * size.getValue());
        pos2.y = pos1.getY();
        pos2.z = pos1.getZ() + (16 * size.getValue());
        sendChatMessage(getChatName() + "Registered chunk: X: [" + pos1.getX() + "][" + pos2.getX() + "] Z: [" + pos1.getZ() + "][" + pos2.getZ() + "]");

        positions = getBlockPositionsInArea(pos1, pos2);
        sendChatMessage(positions.size() + "");
        for (int i = 0 ; i < positions.size() ; i++) { /* Flatten all positions to terrain level */
            positions.set(i, getHighestTerrainPos(positions.get(i)));
            sendChatMessage("Setting " + positions.get(i).getX() + " " + positions.get(i).getZ() + " to " + positions.get(i).getY());
        }
        cornerY[0] = getHighestTerrainPos(pos2).getY();
        cornerY[1] = getHighestTerrainPos(new BlockPos(pos2.getX(), pos1.getY(), pos1.getZ())).getY();
        cornerY[2] = getHighestTerrainPos(new BlockPos(pos1.getX(), pos1.getY(), pos2.getZ())).getY();
        cornerY[3] = getHighestTerrainPos(pos1).getY();
    }

    private boolean insideChunk() {
        return (mc.player.posX > pos1.getX() && pos2.getX() > mc.player.posX) && (mc.player.posZ > pos1.getZ() && pos2.getZ() > mc.player.posZ);
    }

    @Override
    public void onUpdate() {
        if (mc.playerController.getIsHittingBlock()) return;

        if (swing.getValue() && mc.player.ticksExisted % (0.5 * getFrequency()) == 0) {
            Objects.requireNonNull(mc.getConnection()).sendPacket(new CPacketAnimation(EnumHand.MAIN_HAND));
        }

        if (jump.getValue() && mc.player.ticksExisted % (2 * getFrequency()) == 0) {
            mc.player.jump();
        }

        if (mode.getValue().equals(Mode.TRADITIONAL) && turn.getValue() && mc.player.ticksExisted % (0.375 * getFrequency()) == 0) {
            mc.player.rotationYaw = random.nextInt(360) - makeNegRandom(180);
        }
    }

    @Override
    public void onWorldRender(RenderEvent event) {
        if (mc.player == null || positions == null) return;
        GlStateManager.pushMatrix();
        int side = GeometryMasks.Quad.ALL;
        int colourBlue = rgbToInt(155, 144, 255, 50);
        int colourPink = rgbToInt(255, 144, 255, 50);

        KamiTessellator.prepare(GL11.GL_QUADS);

        for (BlockPos renderPos : positions)
            KamiTessellator.drawBox(renderPos, colourBlue, GeometryMasks.Quad.UP);
        KamiTessellator.drawBox(new BlockPos(pos2.getX(), cornerY[0], pos2.getZ()), colourPink, side, cornerY[0] + 2, 1, 1, 0);
        KamiTessellator.drawBox(new BlockPos(pos2.getX(), cornerY[1], pos1.getZ()), colourPink, side, cornerY[1] + 2, 1, 1, 0);
        KamiTessellator.drawBox(new BlockPos(pos1.getX(), cornerY[2], pos2.getZ()), colourPink, side, cornerY[2] + 2, 1, 1, 0);
        KamiTessellator.drawBox(new BlockPos(pos1.getX(), cornerY[3], pos1.getZ()), colourPink, side, cornerY[3] + 2, 1, 1, 0);

        KamiTessellator.release();
        GlStateManager.popMatrix();
        GlStateManager.enableTexture2D();
    }

    @EventHandler
    public Listener<PacketEvent.Receive> receiveListener = new Listener<>(event -> {
        if (autoReply.getValue() && event.getPacket() instanceof SPacketChat && ((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains("whispers: ") && !((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains(mc.player.getName())) {
            sendServerMessage("/r I am currently AFK and using KAMI Blue!");
        }
    });

    private float getFrequency() {
        return reverseNumber(frequency.getValue(), 1, 100);
    }

    private int makeNegRandom(int input) {
        int rand = random.nextBoolean() ? 1 : 0;
        if (rand == 0) {
            return -input;
        } else {
            return input;
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AntiAFK.kt
package me.zeroeightsix.kami.module.modules.misc

import baritone.api.pathing.goals.GoalXZ
import me.zeroeightsix.kami.event.events.BaritoneSettingsInitEvent
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.event.events.SafeTickEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Setting
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.BaritoneUtils
import me.zeroeightsix.kami.util.TimerUtils
import me.zeroeightsix.kami.util.text.MessageDetection
import me.zeroeightsix.kami.util.text.MessageSendHelper.sendServerMessage
import net.minecraft.network.play.server.SPacketChat
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.gameevent.InputEvent
import org.kamiblue.event.listener.listener
import kotlin.random.Random

/**
 * TODO: Path finding to stay inside 1 chunk
 * TODO: Render which chunk is selected
 */
@Module.Info(
    name = "AntiAFK",
    category = Module.Category.MISC,
    description = "Prevents being kicked for AFK"
)
object AntiAFK : Module() {
    private val delay = register(Settings.integerBuilder("ActionDelay").withValue(50).withRange(5, 100).withStep(5))
    private val variation = register(Settings.integerBuilder("Variation").withValue(25).withRange(0, 50))
    private val autoReply = register(Settings.b("AutoReply", true))
    private val swing = register(Settings.b("Swing", true))
    private val jump = register(Settings.b("Jump", true))
    private val turn = register(Settings.b("Turn", true))
    private val walk = register(Settings.b("Walk", true))
    private val radius = register(Settings.integerBuilder("Radius").withValue(64).withRange(1, 128))
    private val inputTimeout = register(Settings.integerBuilder("InputTimeout(m)").withValue(0).withRange(0, 15))

    private var startPos: BlockPos? = null
    private var squareStep = 0
    private var baritoneDisconnectOnArrival = false
    private val inputTimer = TimerUtils.TickTimer(TimerUtils.TimeUnit.MINUTES)
    private val actionTimer = TimerUtils.TickTimer(TimerUtils.TimeUnit.TICKS)
    private var nextActionDelay = 0

    override fun getHudInfo(): String? {
        return if (inputTimeout.value == 0) null
        else ((System.currentTimeMillis() - inputTimer.time) / 1000L).toString()
    }

    override fun onDisable() {
        startPos = null
        BaritoneUtils.settings?.disconnectOnArrival?.value = baritoneDisconnectOnArrival
        BaritoneUtils.cancelEverything()
    }

    override fun onEnable() {
        inputTimer.reset()
        baritoneDisconnectOnArrival()
    }

    init {
        listener<BaritoneSettingsInitEvent> {
            baritoneDisconnectOnArrival()
        }
    }

    private fun baritoneDisconnectOnArrival() {
        BaritoneUtils.settings?.disconnectOnArrival?.let {
            baritoneDisconnectOnArrival = it.value
            it.value = false
        }
    }

    init {
        listener<PacketEvent.Receive> {
            if (!autoReply.value || it.packet !is SPacketChat) return@listener
            if (MessageDetection.Direct.RECEIVE detect it.packet.chatComponent.unformattedText) {
                sendServerMessage("/r I am currently AFK and using KAMI Blue!")
            }
        }

        listener<InputEvent.MouseInputEvent> {
            if (inputTimeout.value != 0 && isInputting()) {
                inputTimer.reset()
            }
        }

        listener<InputEvent.KeyInputEvent> {
            if (inputTimeout.value != 0 && isInputting()) {
                inputTimer.reset()
            }
        }
    }

    private fun isInputting() =
        mc.gameSettings.keyBindAttack.isKeyDown
            || mc.gameSettings.keyBindUseItem.isKeyDown
            || mc.gameSettings.keyBindJump.isKeyDown
            || mc.gameSettings.keyBindSneak.isKeyDown
            || mc.gameSettings.keyBindForward.isKeyDown
            || mc.gameSettings.keyBindBack.isKeyDown
            || mc.gameSettings.keyBindLeft.isKeyDown
            || mc.gameSettings.keyBindRight.isKeyDown

    init {
        listener<SafeTickEvent> {
            if (inputTimeout.value != 0) {
                if (BaritoneUtils.isActive) {
                    inputTimer.reset()
                } else if (!inputTimer.tick(inputTimeout.value.toLong(), false)) {
                    startPos = null
                    return@listener
                }
            }

            if (actionTimer.tick(nextActionDelay.toLong())) {
                val random = if (variation.value > 0) (0..variation.value).random() else 0
                nextActionDelay = delay.value + random

                when ((getAction())) {
                    Action.SWING -> mc.player.swingArm(EnumHand.MAIN_HAND)
                    Action.JUMP -> mc.player.jump()
                    Action.TURN -> mc.player.rotationYaw = Random.nextDouble(-180.0, 180.0).toFloat()
                }

                if (walk.value && !BaritoneUtils.isActive) {
                    squareWalk()
                }
            }
        }
    }

    private fun getAction(): Action? {
        if (!swing.value && !jump.value && !turn.value) return null
        val action = Action.values().random()
        return if (action.setting.value) action else getAction()
    }

    private fun squareWalk() {
        if (startPos == null) startPos = mc.player.position

        startPos?.let {
            when (squareStep) {
                0 -> baritoneGotoXZ(it.x, it.z + radius.value)
                1 -> baritoneGotoXZ(it.x + radius.value, it.z + radius.value)
                2 -> baritoneGotoXZ(it.x + radius.value, it.z)
                3 -> baritoneGotoXZ(it.x, it.z)
            }
        }

        squareStep = (squareStep + 1) % 4
    }

    private fun baritoneGotoXZ(x: Int, z: Int) {
        BaritoneUtils.primary?.customGoalProcess?.setGoalAndPath(GoalXZ(x, z))
    }

    private enum class Action(val setting: Setting<Boolean>) {
        SWING(swing),
        JUMP(jump),
        TURN(turn)
    }

    init {
        walk.settingListener = Setting.SettingListeners {
            BaritoneUtils.cancelEverything()
        }
    }
}